### PR TITLE
ChainSyncClient: include the reason why a block is invalid

### PR DESCRIPTION
**Pull request by mrBliss**
Addresses #1254.

The ChainSyncClient will disconnect from a peer when it sees a header in the
candidate chain that it knows corresponds to an invalid block. How does it
know it is invalid? The block has been downloaded in the past, the ChainDB has
validated the corresponding block, found it to be invalid, traced this along
with the reason why it was invalid, and recorded that the block is invalid.

However, many people are wondering why the ChainSyncClient thinks a block is
invalid, as it doesn't include the reason in its exception/trace message. Note
that the ChainDB **must** have already traced the reason why that block was
invalid, but that trace was maybe not enabled or people don't bother to scroll
up.

So to make all our lives easier, we now remember the exact reason why a block
is invalid in the ChainDB, instead of merely remembering the fact that a block
was invalid. This allows us to include the exact reason in the exception/trace
message why a ChainSyncClient thinks a block is invalid.

### CODE REVIEW

**Comment by mrBliss** [comment]
This PR also makes it possible to have better trace messages in #787 and #1225.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron/Orphans.hs [hunk 1]
```diff
@@ -123,3 +130,16 @@ instance Condense (CC.AMempoolPayload a) where
...
+-------------------------------------------------------------------------------}
+
+deriving instance Generic CC.ChainValidationError
+
+-- TODO a proper instance.
+--
+-- Problem: Cardano.Chain.Delegation.Validation.Registration.TooLarge is not
+-- exported, but occurs somewhere in CC.ChainValidationError.
+deriving via (OnlyCheckIsWHNF "ChainValidationError" CC.ChainValidationError)
+  instance NoUnexpectedThunks CC.ChainValidationError
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron/Orphans.hs]: https://github.com/input-output-hk/cardano-ledger/issues/683

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron/Orphans.hs]: Use `UseIsNormalForm`.

**Pull request merged**